gcsfs/core.py

```diff
@@ -217,87 +223,123 @@ def load_tokens():
+        self.tokens[(project, access)] = self.token
+        self._save_tokens()
+
+    def connect(self, method=None):
         """
         Establish session token. A new token will be requested if the current
         one is within 100s of expiry.
 
         Parameters
         ----------
-        refresh: bool (False)
-            Force refresh, even if the token is expired.
+        method: str (gtoken|cache|cloud|token|anon|browser) or None
+            Type of authorisation to implement - calls `_connect_*` methods.
+            If None, will try sequence of methods.
         """
-        token = self.input_token
         self.session = requests.Session()
-        project, access = self.project, self.access
-        if token is not None and str(token) != 'cloud':
-            if 'type' in token or isinstance(token, str):
-                token = self._parse_gtoken(token)
-            self.tokens[(project, access)] = token
-        if (project, access) in self.tokens:
-            # cached credentials
-            data = self.tokens[(project, access)]
-        elif token == 'cloud':
-            data = {'timestamp': time.time() - 3600, 'expires_in': 1,
-                    'type': 'cloud'}
-        else:
-            try:
-                data = self.get_default_gtoken()
-                self.tokens[(project, access)] = data
-            except:
-                # no credentials - try to ask google in the browser
-                scope = "https://www.googleapis.com/auth/devstorage." + access
-                path = 'https://accounts.google.com/o/oauth2/device/code'
-                r = self.session.post(
-                    path, params={'client_id': not_secret['client_id'],
-                                  'scope': scope})
-                validate_response(r, path)
-                data = json.loads(r.content.decode())
-                print('Navigate to:', data['verification_url'])
-                print('Enter code:', data['user_code'])
-                while True:
-                    time.sleep(1)
-                    r = self.session.post(
-                            "https://www.googleapis.com/oauth2/v4/token",
-                            params={
-                                'client_id': not_secret['client_id'],
-                                'client_secret': not_secret['client_secret'],
-                                'code': data['device_code'],
-                                'grant_type':
-                                    "http://oauth.net/grant_type/device/1.0"})
-                    data2 = json.loads(r.content.decode())
-                    if 'error' in data2:
-                        continue
-                    data = data2
+        if method not in ['gtoken', 'cache', 'cloud', 'token', 'anon',
+                          'browser', None]:
+            method = 'token'
+        if method is None:
+            for meth in ['gtoken', 'cache', 'cloud', 'anon']:
+                self.connect(method=meth)
```

Comment:
Perhaps this should return success or failure?  Perhaps it should raise and we should try-except around it?

Reply:
Basically, if `self.token` remains unset (None), auth failed for that method. I had tried return True/False, but it felt pretty ugly. The exception is `'anon'`, so that could instead give `self.token=False` or other special value.